readthedocs/projects/tasks.py

```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
             self.config = None
 
             setup_successful = self.run_setup(record=record)
-            if setup_successful:
-                self.run_build(record=record, docker=docker)
+            if not setup_successful:
+                return False
+        # Catch unhandled errors in the setup step
         except Exception as e:  # noqa
             log.exception(
-                'An unhandled exception was raised outside the build environment',
+                'An unhandled exception was raised during build setup',
                 extra={'tags': {'build': build_pk}}
             )
-            unhandled_failure = _(
-                'Unknown error encountered. '
+            self.setup_env.build['error'] = _(
+                'An unknown error was encountered while setting up your project. '
                 'Please include the build id ({build_id}) in any bug reports.'.format(
                     build_id=build_pk
                 ))
-        finally:
-            if unhandled_failure:
-                self.build_env.build['error'] = unhandled_failure
-            self.build_env.update_build(BUILD_STATE_FINISHED)
+            self.setup_env.update_build(BUILD_STATE_FINISHED)
+            return False
+        else:
+            # No exceptions in the setup step, catch unhandled errors in the
+            # build steps
+            try:
+                self.run_build(record=record, docker=docker)
+            except Exception as e:  # noqa
+                log.exception(
+                    'An unhandled exception was raised during project build',
+                    extra={'tags': {'build': build_pk}}
+                )
+                self.build_env.build['error'] = _(
+                    'An unknown error was encountered while building your project. '
+                    'Please include the build id ({build_id}) in any bug reports.'.format(
+                        build_id=build_pk
+                    ))
+                self.build_env.update_build(BUILD_STATE_FINISHED)
```

Comment:
Another place we're updating build finished..?

Reply:
It seems this is what we want, no? This is the fallback that was added in case there are any exceptions outside of the build environment `__enter__`, normal context, or in `__exit__`, such as in `run_build`.

Reply:
Ah, ok. I got lost when reviewing this for some reason, and didn't realize where everything was getting called from.